File: src/main/java/com/hoi4utils/clauzewitz/code/effect/Parameter.java
package main.java.com.hoi4utils.clauzewitz.code.effect;

import main.java.com.hoi4utils.clausewitz_parser.NodeValue;
import main.java.com.hoi4utils.clauzewitz.exceptions.NullParameterTypeException;

import java.util.*;

public class Parameter implements EffectParameter, Cloneable {
	private static final HashMap<String, Parameter> allParameters = new HashMap<>();
	// private static final HashSet<Parameter> allValidParameters = new HashSet<>();
	// static {
	// Parameter p = new Parameter(null, ParameterValueType.);
	// allValidParameters.add(p);
	// }
	final List<ParameterValueType> allowedParameterValueTypes;

	final String identifier;
	ParameterValueType parameterValueType; // todo necessary?
	ParameterValue value;

	Parameter(String name, List<ParameterValueType> allowedParameterValueTypes) throws NullParameterTypeException {
		this.identifier = name;
		if (allowedParameterValueTypes == null) {
			throw new NullParameterTypeException("allowedParameterValueType null, " + name);
		} else {
			this.allowedParameterValueTypes = allowedParameterValueTypes;
		}
		allParameters.put(name, this);
	}

	Parameter(String name, ParameterValueType allowedParameterValueType) throws NullParameterTypeException {
		this.identifier = name;
		if (allowedParameterValueType == null) {
			throw new NullParameterTypeException("allowedParameterValueType null, " + name);
		} else {
			this.allowedParameterValueTypes = List.of(allowedParameterValueType);
		}
		allParameters.put(name, this);
	}

	Parameter(String name) throws NullParameterTypeException {
		this(name, (ParameterValueType) null);
	}

	public static Parameter of(String name, NodeValue v) {
		Parameter parameter = getClone(name);
		if (parameter == null)
			return null;
		parameter.value = new ParameterValue(v);
		return parameter;
	}

	public static Parameter of(NodeValue v) {
		return of(null, v);
	}

	private static Parameter getClone(String name) {
		Parameter clone;
		try {
			Parameter parameter = allParameters.get(name);
			if (parameter == null)
				return null;
			clone = (Parameter) parameter.clone();
		} catch (CloneNotSupportedException e) {
			throw new RuntimeException(e);
		}
		return clone;
	}

	public static boolean containsScopeParameter(List<Parameter> requiredParameters) {
		for (Parameter p : requiredParameters) {
			if (p.allowedParameterValueTypes.contains((ParameterValueType.scope))) {
				return true;
			}
		}
		return false;
	}

	@Override
	protected Object clone() throws CloneNotSupportedException {
		Parameter c = (Parameter) super.clone();
		c.value = null;
		return c;
	}

	// public static void addValidParameter(String identifier,
	// Set<ParameterValueType> validParameterValueTypes) {
	//// if (!allParameters.keySet().contains(identifier)) {
	//// Parameter p = new Parameter(identifier, validParameterValueTypes);
	//// allParameters.put(identifier, p);
	//// } else {
	//// allParameters.get(identifier).addValidParameterValueTypes(validParameterValueTypes);
	//// }
	// }

	public static boolean isParameter(String name, NodeValue v) {
		if (v.isString()) {
			return allParameters.containsKey(v.string());
		}
		if (v.isNumber()) {
			return true;
		}
		if (v.isList()) {
			// todo not sure
			return true; // maybe? hm..
		}

		return false;
	}

	@Override
	public String displayScript() {
		// return "[parameter]";
		if (identifier == null) {
			return value.toString();
		}
		return identifier + " = " + value.toString();
	}

	private static class ParameterValue {
		public Object value;

		public ParameterValue(String value) {
			this.value = value;
		}

		public ParameterValue(NodeValue value) {
			if (!value.isList()) {
				this.value = value.valueObject();
			} else {
				System.err.println("effect value can not be list.");
			}
		}

		@Override
		public String toString() {
			StringBuilder s = new StringBuilder();
			if (value == null) {
				return "[null parameter value]";
			}

			if (value instanceof List) {
				// s.append(identifier);
				// s.append(" = ");
				// s.append("{\n");
				// for (var n : (List<Node>) value) {
				// s.append("\t");
				// s.append(n.);
				// s.append("\n");
				// }
				// s.append("}\n");
				s.append("[parameter -> list]");
			} else {
				s.append(value.toString());
			}
			return s.toString();
		}
	}

}

File: _src_old/main/scala/com/hoi4utils/databases/effect/Parameter.scala
package com.hoi4utils.databases.effect

import com.hoi4utils.exceptions.NullParameterTypeException

import scala.collection.mutable

object Parameter {
  private val allParameters = mutable.HashMap.empty[String, Parameter]

  /** 
   * Create or register a parameter that allows exactly these types. 
   **/
  def apply(name: String, allowedTypes: List[ParameterValueType]): Parameter =
    new Parameter(name, allowedTypes)

  /** 
   * Create or register a parameter that allows a single type. 
   **/
  def apply(name: String, allowedType: ParameterValueType): Parameter =
    new Parameter(name, List(allowedType))

  /** 
   * Create or register a parameter with no explicit allowed types.
   */
  def apply(name: String): Parameter =
    new Parameter(name, Nil)

  /** 
   * Clone an existing parameter by name, clearing its runtime value.
   */
  private def getClone(name: String): Option[Parameter] =
    allParameters.get(name).map(_.cloneParam())

  /** 
   * Returns true if any of the given parameters accepts `Scope` as a value type. 
   */
  def containsScopeParameter(params: List[Parameter]): Boolean =
    params.exists(_.allowedParameterValueTypes.contains(ParameterValueType.scope))

  // Internal wrapper for the value you assign at runtime
  private case class ParameterValue(value: AnyRef) {
    override def toString: String =
      value match {
        case null        => "[null parameter value]"
        case l: List[_]  => "[parameter -> list]"
        case other       => other.toString
      }
  }
}

class Parameter private (val identifier: String, var allowedParameterValueTypes: List[ParameterValueType]) 
  extends EffectParameter with Cloneable {

  import Parameter.ParameterValue

  // register in the global map
  if (identifier == null)
    throw NullParameterTypeException("Parameter identifier cannot be null")
  Parameter.allParameters.put(identifier, this)

  private var runtimeValue: Option[ParameterValue] = None

  /** 
   * Set a runtime value for this parameter (e.g. when you call `of(name, value)`). 
   */
  def withValue(v: AnyRef): this.type = {
    runtimeValue = Some(ParameterValue(v))
    this
  }

  /** 
   * Clone this parameter (clearing any runtime value). 
   */
  def cloneParam(): Parameter = {
    val c = this.clone().asInstanceOf[Parameter]
    c.runtimeValue = None
    c
  }

  @throws[CloneNotSupportedException]
  override protected def clone(): AnyRef =
    super.clone()

  /** 
   * Render as PDX script: "id = value" or just the value if no id 
   */
  def displayScript: String =
    if (identifier.nonEmpty) s"$identifier = ${runtimeValue.map(_.toString).getOrElse("")}"
    else runtimeValue.map(_.toString).getOrElse("")

  override def toString: String = displayScript
}
